File: distribution/python/dqops/client/models/check_spec_model.py
from typing import Any, Dict, List, Type, TypeVar, Union

import attr

from ..types import UNSET, Unset

T = TypeVar("T", bound="CheckSpecModel")


@attr.s(auto_attribs=True)
class CheckSpecModel:
    """Check spec model

    Attributes:
        check_name (Union[Unset, str]): Check name
        sensor_name (Union[Unset, str]): Sensor name
        rule_name (Union[Unset, str]): Rule name
        help_text (Union[Unset, str]): Help text that is shown in the check editor that describes the purpose and usage
            of the check
        custom (Union[Unset, bool]): This check has is a custom check or was customized by the user.
        built_in (Union[Unset, bool]): This check is provided with DQO as a built-in check.
    """

    check_name: Union[Unset, str] = UNSET
    sensor_name: Union[Unset, str] = UNSET
    rule_name: Union[Unset, str] = UNSET
    help_text: Union[Unset, str] = UNSET
    custom: Union[Unset, bool] = UNSET
    built_in: Union[Unset, bool] = UNSET
    additional_properties: Dict[str, Any] = attr.ib(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        check_name = self.check_name
        sensor_name = self.sensor_name
        rule_name = self.rule_name
        help_text = self.help_text
        custom = self.custom
        built_in = self.built_in

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if check_name is not UNSET:
            field_dict["check_name"] = check_name
        if sensor_name is not UNSET:
            field_dict["sensor_name"] = sensor_name
        if rule_name is not UNSET:
            field_dict["rule_name"] = rule_name
        if help_text is not UNSET:
            field_dict["help_text"] = help_text
        if custom is not UNSET:
            field_dict["custom"] = custom
        if built_in is not UNSET:
            field_dict["built_in"] = built_in

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        check_name = d.pop("check_name", UNSET)

        sensor_name = d.pop("sensor_name", UNSET)

        rule_name = d.pop("rule_name", UNSET)

        help_text = d.pop("help_text", UNSET)

        custom = d.pop("custom", UNSET)

        built_in = d.pop("built_in", UNSET)

        check_spec_model = cls(
            check_name=check_name,
            sensor_name=sensor_name,
            rule_name=rule_name,
            help_text=help_text,
            custom=custom,
            built_in=built_in,
        )

        check_spec_model.additional_properties = d
        return check_spec_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/rest/models/metadata/CheckSpecModel.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.rest.models.metadata;

import com.dqops.metadata.definitions.checks.CheckDefinitionSpec;
import com.dqops.metadata.definitions.checks.CheckDefinitionWrapper;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.Objects;

/**
 * CheckSpec model that is returned by the REST API. Describes a single unique data quality check.
 */
@Data
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@ApiModel(value = "CheckSpecModel", description = "Check spec model")
public class CheckSpecModel {
    /**
     * Data quality check name.
     */
    @JsonPropertyDescription("Check name")
    private String checkName;

    /**
     * Full sensor name.
     */
    @JsonPropertyDescription("Sensor name")
    private String sensorName;

    /**
     * Full rule name.
     */
    @JsonPropertyDescription("Rule name")
    private String ruleName;

    /**
     * Help text for the check editor.
     */
    @JsonPropertyDescription("Help text that is shown in the check editor that describes the purpose and usage of the check")
    private String helpText;

    /**
     * True when the check is a custom check or is customized by the user.
     */
    @JsonPropertyDescription("This check has is a custom check or was customized by the user.")
    private boolean custom;

    /**
     * True when this check is provided with DQO as a built-in check.
     */
    @JsonPropertyDescription("This check is provided with DQO as a built-in check.")
    private boolean builtIn;

    /**
     * Boolean flag that decides if the current user can update or delete this object.
     */
    @JsonPropertyDescription("Boolean flag that decides if the current user can update or delete this object.")
    private boolean canEdit;

    /**
     * Default constructor for CheckModel.
     */
    public CheckSpecModel() {
    }

    /**
     * Constructor for a rule model with a given {@link CheckDefinitionWrapper}, custom flag, and built-in flag.
     * Sets the properties of the check based on the given {@link CheckDefinitionWrapper}.
     * @param checkDefinitionWrapper The check definition wrapper to use.
     * @param custom Whether the check is a custom check.
     * @param builtIn Whether the check is a built-in check.
     * @param canEdit The current user can edit the check definition.
     */
    public CheckSpecModel(CheckDefinitionWrapper checkDefinitionWrapper, boolean custom, boolean builtIn, boolean canEdit) {
        this.checkName = checkDefinitionWrapper.getCheckName();
        this.sensorName = checkDefinitionWrapper.getSpec().getSensorName();
        this.ruleName = checkDefinitionWrapper.getSpec().getRuleName();
        this.helpText = checkDefinitionWrapper.getSpec().getHelpText();
        this.custom = custom;
        this.builtIn = builtIn;
        this.canEdit = canEdit;
    }

    /**
     * Checks whether the CheckModel is equal to a given {@link CheckDefinitionWrapper}.
     * @param checkDefinitionWrapper The CheckDefinitionWrapper to compare.
     * @return true if the rule model is equal to the CheckDefinitionWrapper, false otherwise.
     */
    public boolean equalsBuiltInCheck(CheckDefinitionWrapper checkDefinitionWrapper) {
        if (checkDefinitionWrapper == null) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getSensorName(), this.sensorName)) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getRuleName(), this.ruleName)) {
            return false;
        }

        if (!Objects.equals(checkDefinitionWrapper.getSpec().getHelpText(), this.helpText)) {
            return false;
        }

        return true;
    }

    /**
     * Returns the check definition spec object.
     * @return the check definition spec object.
     */
    public CheckDefinitionSpec toCheckDefinitionSpec() {
        CheckDefinitionSpec checkDefinitionSpec = new CheckDefinitionSpec();
        checkDefinitionSpec.setSensorName(this.sensorName);
        checkDefinitionSpec.setRuleName(this.ruleName);
        checkDefinitionSpec.setHelpText(this.helpText);

        return checkDefinitionSpec;
    }
}
